File: frontend/scripts/check-i18n.mjs
// Every t("…") key a component uses must exist in en.json, and every other bundle
// must not drift into keys English does not have. A missing key renders as the raw
// key string in the UI, which no typechecker catches.
import { readFileSync, readdirSync, statSync } from "node:fs";
import { join } from "node:path";

const SRC = "src";
const files = [];
(function walk(dir) {
  for (const e of readdirSync(dir)) {
    const p = join(dir, e);
    if (statSync(p).isDirectory()) walk(p);
    else if (/\.tsx?$/.test(p)) files.push(p);
  }
})(SRC);

const en = JSON.parse(readFileSync(join(SRC, "i18n/en.json"), "utf8"));
const used = new Map(); // key -> file
const dynamic = [];

for (const f of files) {
  const text = readFileSync(f, "utf8");
  // t("literal.key") and t(`literal.${expr}`) — the latter is reported, not resolved.
  for (const m of text.matchAll(/\bt\(\s*"([^"]+)"/g)) used.set(m[1], f);
  for (const m of text.matchAll(/\bt\(\s*`([^`]*\$\{[^`]*)`/g)) dynamic.push([m[1], f]);
  // labelKey fields hold keys resolved through t() elsewhere.
  for (const m of text.matchAll(/labelKey:\s*"([^"]+)"/g)) used.set(m[1], f);
}

let bad = 0;
for (const [key, file] of [...used].sort()) {
  if (!(key in en)) {
    console.error(`missing in en.json: ${key}   (${file})`);
    bad++;
  }
}

// Template keys: check every prefix has at least one matching key, which catches a
// whole family being absent (e.g. no device.confidence.* at all).
for (const [tpl, file] of dynamic) {
  const prefix = tpl.split("${")[0];
  if (prefix && !Object.keys(en).some((k) => k.startsWith(prefix))) {
    console.error(`no key starts with "${prefix}" (from \`${tpl}\`)   (${file})`);
    bad++;
  }
}

for (const lang of ["de", "fr", "it"]) {
  const b = JSON.parse(readFileSync(join(SRC, `i18n/${lang}.json`), "utf8"));
  for (const k of Object.keys(b)) {
    if (!(k in en)) {
      console.error(`${lang}.json has a key en.json lacks: ${k}`);
      bad++;
    }
  }
  // A missing key is an *error*, not a warning. English is the fallback (FR-4), so a
  // missing German key silently renders the English string and a half-translated screen
  // looks finished to anybody testing in English. Warning about it meant nothing noticed
  // for as long as this file was not run in CI, which was its whole first year.
  const missing = Object.keys(en).filter((k) => !(k in b));
  if (missing.length) {
    console.error(
      `${lang}.json is missing ${missing.length} key(s), which will silently render in ` +
        `English:\n    ${missing.join("\n    ")}`,
    );
    bad += missing.length;
  }
}

// Keys in the bundle that nothing reads. An error rather than a warning: eight had
// accumulated behind a console.warn, including three left over from a screen that was
// redesigned. Values reached through a variable are counted as used -- the Data screen's
// SOURCES table holds one key per source -- so this is safe to fail on.
const referenced = new Set();
for (const f of files) {
  for (const m of readFileSync(f, "utf8").matchAll(/["']([a-z][A-Za-z0-9_]*(?:\.[A-Za-z0-9_]+)+)["']/g)) {
    referenced.add(m[1]);
  }
}
const prefixes = dynamic.map(([tpl]) => tpl.split("${")[0]).filter(Boolean);
const dead = Object.keys(en).filter(
  (k) => !used.has(k) && !referenced.has(k) && !prefixes.some((p) => k.startsWith(p)),
);
if (dead.length) {
  console.error(`${dead.length} key(s) defined but never used:\n    ${dead.join("\n    ")}`);
  bad += dead.length;
}

if (bad) {
  console.error(`\n${bad} problem(s)`);
  process.exit(1);
}
console.log(`i18n ok: ${used.size} literal keys, ${dynamic.length} template keys, ${Object.keys(en).length} in en.json`);

File: tools/check_i18n.py
#!/usr/bin/env python3
"""Check the four language bundles against each other and against the components.

SPEC.md FR-4 requires de / fr / it / en with English as the fallback. The fallback is
what makes an incomplete bundle invisible: a missing German key silently renders the
English string, so a half-translated screen looks finished to anybody testing in
English. This is the only thing that catches it.

Three checks:

1. Every bundle has exactly the keys English has. A missing key falls back silently; an
   extra key is dead weight that suggests a rename went half-done.
2. Every literal key used in a component exists. A missing one renders as its own key
   name -- "about.trademarks" in the middle of a page -- which reviewers do notice, but
   only if they open that screen.
3. Every key in the bundle is used. Reported as a warning, not an error, because keys
   are also built from template literals (`preset.${id}`), and those prefixes are
   collected and treated as covering everything beneath them.

Run: python3 tools/check_i18n.py
"""

import json
import pathlib
import re
import sys

ROOT = pathlib.Path(__file__).resolve().parent.parent
I18N = ROOT / "frontend/src/i18n"
SRC = ROOT / "frontend/src"
LANGS = ["en", "de", "fr", "it"]

# t("literal.key") and t('literal.key')
LITERAL = re.compile(r"""\bt\(\s*["']([A-Za-z0-9_.]+)["']""")
# t(`prefix.${expr}`) -- the prefix covers every key under it
TEMPLATE = re.compile(r"""\bt\(\s*`([A-Za-z0-9_.]*?)\$\{""")
# Any dotted string literal. Some keys reach `t` through a variable rather than
# directly -- the Data screen's SOURCES table stores one per source -- and counting
# only direct calls reported twenty of those as unused.
REFERENCED = re.compile(r"""["']([a-z][A-Za-z0-9_]*(?:\.[A-Za-z0-9_]+)+)["']""")


def main() -> int:
    bundles = {lang: json.loads((I18N / f"{lang}.json").read_text()) for lang in LANGS}
    english = set(bundles["en"])
    problems = []

    for lang in LANGS[1:]:
        keys = set(bundles[lang])
        missing = sorted(english - keys)
        extra = sorted(keys - english)
        if missing:
            problems.append(
                f"{lang}.json is missing {len(missing)} key(s) English has, so they "
                f"fall back to English silently:\n    " + "\n    ".join(missing)
            )
        if extra:
            problems.append(
                f"{lang}.json has {len(extra)} key(s) English does not:\n    "
                + "\n    ".join(extra)
            )

    used, prefixes, referenced = set(), set(), set()
    for path in SRC.rglob("*.ts*"):
        text = path.read_text()
        used |= set(LITERAL.findall(text))
        prefixes |= set(TEMPLATE.findall(text))
        referenced |= set(REFERENCED.findall(text))

    undefined = sorted(k for k in used if k not in english)
    if undefined:
        problems.append(
            f"{len(undefined)} key(s) used in components but defined nowhere, which "
            f"render as their own name:\n    " + "\n    ".join(undefined)
        )

    unused = sorted(
        k
        for k in english
        if k not in used
        and k not in referenced
        and not any(k.startswith(p) for p in prefixes if p)
    )

    for p in problems:
        print(f"ERROR: {p}", file=sys.stderr)
    if unused:
        print(f"warning: {len(unused)} key(s) defined but never used: {', '.join(unused)}")

    if problems:
        return 1
    print(
        f"OK: {len(english)} keys x {len(LANGS)} languages, "
        f"{len(used)} literal uses, {len(prefixes)} template prefixes"
    )
    return 0


if __name__ == "__main__":
    sys.exit(main())
